DDOT_module/parent_classes/marginals_analyzer_base.py

```diff
@@ -0,0 +1,297 @@
+
+    @marginals_dictionary.setter
+    def marginals_dictionary(self, marginals_dictionary: Dict[str, Dict[str, np.ndarray]]) -> None:
+        self._marginals_dictionary = marginals_dictionary
+
+    @staticmethod
+    def get_qubits_key(list_of_qubits: List[int]) -> str:
+        """ from subset of qubit indices get the string that labels this subset
+            using convention 'q5q6q12...' etc.
+        :param list_of_qubits: labels of qubits
+
+        :return: string label for qubits
+
+         NOTE: this function is "dual" to self.get_qubit_indices_from_string.
+        """
+
+        return 'q' + 'q'.join([str(s) for s in list_of_qubits])
+
+    @staticmethod
+    def get_qubit_indices_from_string(qubits_string: str,
+                                      with_q: Optional[bool] = False):
+
+        """Return list of qubit indices from the string of the form "q0q1q22q31"
+        :param qubits_string: string which has the form of "q" followed by qubit index
+        :param with_q: specify whether returned indices should be in form of string with letter
+
+        :return: list of qubit indices:
+
+        depending on value of parameter "with_q" the mapping will be one of the following:
+        if with_q:
+            'q1q5q13' -> ['q1','q5','q13']
+        else:
+            'q1q5q13' -> [1,5,13]
+
+        NOTE: this function is "dual" to self.get_qubits_key.
+        """
+
+        numbers = re.findall(r'\d+', qubits_string)
+
+        if with_q:
+            qubits = ['q' + s for s in numbers]
+        else:
+            qubits = [int(s) for s in numbers]
+
+        return qubits
+
+    def results_dictionary_update(self,
+                                  results_dictionary_new: Dict[str, Dict[str, int]]) -> None:
+        # This method updates results dictionary from class property with new dictionary.
+        # Note that if there is KEY collision, then the value from new dictionary overwrites old one.
+
+        self._results_dictionary = {**self._results_dictionary,
+                                    **results_dictionary_new}
+
+    def marginals_dictionary_update(self,
+                                    marginals_dictionary_new: Dict[
+                                        str, Dict[str, np.ndarray]]) -> None:
+        # See description of self.results_dictionary_update
+
+        self._marginals_dictionary = {**self._marginals_dictionary,
+                                      **marginals_dictionary_new}
+
+    def normalize_marginals(self):
+        # Loop through all experiments and marginals and normalize them.
+        for key_experiment in self._marginals_dictionary.keys():
+            for key_marginal in self._marginals_dictionary[key_experiment].keys():
+                self._marginals_dictionary[key_experiment][key_marginal] *= 1 / np.sum(
+                    self._marginals_dictionary[key_experiment][key_marginal])
+
+    def compute_marginals(self,
+                          experiment_key: str,
```

Comment:
Here the argument is named `experiment_key` whereas in one of the methods above you've used variable with the name `key_experiment` (for the same type of variable, I believe). Be consistent.

I like the first name better if that's of any value for you. :P